docs/LoongArch-ELF-ABI-EN.adoc

```diff
@@ -628,3 +628,197 @@ with check 32-bit unsigned overflow
+If no GAR is available, it’s passed on the stack.
+When passed in registers or on the stack, integer scalars narrower than GRLEN bits are sign-extended to GRLEN bits.
+
+. GRLEN < WOA ≤ 2 × GRLEN
+
+.. The argument is passed in a pair of GAR, with the low-order GRLEN bits in the lower-numbered register and the high-order GRLEN bits in the higher-numbered register.
+If exactly one register is available, the low-order GRLEN bits are passed in the register and the high-order GRLEN bits are passed on the stack.
+If no GAR is available, it’s passed on the stack.
+
+. WOA > 2 × GRLEN
+
+.. Scalars are passed by reference and are replaced in the argument list with the address.
+
+==== Structure
+
+Empty structures are ignored by C compilers which support them as a non-standard extension(same as union arguments and return values).
+Bits unused due to padding, and bits past the end of a structure whose size in bits is not divisible by GRLEN, are undefined.
+And the layout of the structure on the stack is consistent with that in memory.
+
+. 0 < WOA ≤ GRLEN
+
+.. The structure has only fixed-point members.
+If there is an available GAR, the structure is passed through the GAR by value passing; If no GAR is available, it’s passed on the stack.
+
+.. The structure has only float-point members:
+
+... One double member.
+The argument is passed in a FAR; If no FAR is available, the value is passed in a GAR; If there is no available GAR, the value is passed on the stack.
+
+... Two float members.
+The argument is passed in a pair of available FAR, with the low-order float member bits in the lower-numbered FAR and the high-order float member bits in the higher-numbered FAR.
+If the number of available FAR is less than 1, it’s passed in a GAR, and passed on the stack if there is no available GAR.
+
+.. The structure has both fixed-point and floating-point members, i.e. the structure has one float member and
+
+... Multiple fixed-point members.
+If there are available GAR, the structure is passed in a GAR, and passed on the stack if there is no available GAR.
+
+... Only one fixed-point member.
+If one FAR and one GAR are available, the floating-point member of the structure is passed in the FAR, and the integer member of the structure is passed in the GAR; If no floating-point register but one GAR is available, it’s passed in GAR; If no GAR is available, it’s passed on the stack.
+
+. GRLEN < WOA ≤ 2 × GRLEN
+
+.. Only fixed-point members.
+
+... The argument is passed in a pair of available GAR, with the low-order bits in the lower-numbered GAR and the high-order bits in the higher-numbered GAR.
+If the number of available GAR is 1, the low-order bits are in the GAR and the high-order bits are on the stack, and passed on the stack if there is no available GAR.
+
+.. Only float-point members.
+
+... i.e. The structure has one long double member or one double member and two adjacent float members or 3-4 float members.
```

Comment:
Does this also apply to a structure that has two `double` members? Is structure with two `double` members passed in GARs?

Reply:
> Does this also apply to a structure that has two `double` members? Is structure with two `double` members passed in GARs?

Yes, the struct is passed by two integer register(GARs).

By the way, I had asked the author that `... i.e. The structure has one long double member or one double member and two adjacent float members or 3-4 float members.`
He wanted to express that:
struct {long double a;} or   struct {double a; double b;}  or struct {double a; float b; float c;} or struct {float a; float b; float c; float d;}.

Reply:
So the only case of GRLEN < WOA ≤ 2 × GRLEN structures that is candidate for FAR is one double member and one fixed-point member. All of other cases are passed in GAR. I think that the description can be simplified and just mention this one case as an exception.

Is there a rationale for why the one double member and one fixed-point member structures are special cased like this? It sounds like a lot of complexity for very little gain.

Reply:
> So the only case of GRLEN < WOA ≤ 2 × GRLEN structures that is candidate for FAR is one double member and one fixed-point member. All of other cases are passed in GAR. I think that the description can be simplified and just mention this one case as an exception.

Yes, there is room for improment. 

> Is there a rationale for why the one double member and one fixed-point member structures are special cased like this? It sounds like a lot of complexity for very little gain.

Now it's too late for amendment of the ABI. (By the way, the RISC-V has the same situation)
This ABI for gcc/llvm might be unconspicuous, but for JIT is a lot fo comlexity and maybe worse for gain.

For dotnet-runtime, I have to implement the ABI following this spec only.

Reply:
> Does this also apply to a structure that has two `double` members? Is structure with two `double` members passed in GARs?

Actually, a structure with two 'double' members passed in pair of available FARS.

Reply:
> Does this also apply to a structure that has two `double` members? Is structure with two `double` members passed in GARs?

Sorry for missing this. I've added it in new submit.